Sources/SwiftSyntaxMacroExpansion/MacroReplacement.swift

```diff
@@ -254,31 +325,72 @@ private final class MacroExpansionRewriter: SyntaxRewriter {
     // Swap in the argument for this parameter
     return arguments[parameterIndex].trimmed
   }
+
+  override func visit(_ node: GenericArgumentSyntax) -> GenericArgumentSyntax {
+    guard let parameterIndex = genericParameterReplacements[node] else {
+      return super.visit(node)
+    }
+
+    guard parameterIndex < genericArguments.count else {
+      return super.visit(node)
+    }
+
+    // Swap in the argument for type parameter
+    return GenericArgumentSyntax(
+      leadingTrivia: node.leadingTrivia,
+      node.unexpectedBeforeArgument,
+      argument: genericArguments[parameterIndex].trimmed,
+      node.unexpectedBetweenArgumentAndTrailingComma,
+      trailingComma: node.trailingComma,
+      node.unexpectedAfterTrailingComma
+        // TODO: seems we're getting spurious trailing " " here,
+        //  skipping trailing trivia for now
+        // trailingTrivia: node.trailingTrivia
+    )
+  }
 }
 
 extension MacroDeclSyntax {
   /// Expand the definition of this macro when provided with the given
   /// argument list.
   private func expand(
     argumentList: LabeledExprListSyntax?,
+    genericArgumentList: GenericArgumentClauseSyntax?,
     definition: MacroExpansionExprSyntax,
-    replacements: [MacroDefinition.Replacement]
+    replacements: [MacroDefinition.Replacement],
+    genericReplacements: [MacroDefinition.GenericArgumentReplacement] = []
   ) -> ExprSyntax {
     // FIXME: Do real call-argument matching between the argument list and the
     // macro parameter list, porting over from the compiler.
+    let parameterReplacements = Dictionary(
+      replacements.map { replacement in
+        (replacement.reference, replacement.parameterIndex)
+      },
+      uniquingKeysWith: { l, r in l }
+    )
     let arguments: [ExprSyntax] =
       argumentList?.map { element in
         element.expression
       } ?? []
 
-    return MacroExpansionRewriter(
-      parameterReplacements: Dictionary(
-        uniqueKeysWithValues: replacements.map { replacement in
-          (replacement.reference, replacement.parameterIndex)
-        }
-      ),
-      arguments: arguments
-    ).visit(definition)
+    let genericReplacements = Dictionary(
+      genericReplacements.map { replacement in
+        (replacement.reference, replacement.parameterIndex)
+      },
+      uniquingKeysWith: { l, r in l }
+    )
+    let genericArguments: [TypeSyntax] =
+      genericArgumentList?.arguments.map { element in
+        element.argument
+      } ?? []
```

Comment:
Could be shortened slightly to

```suggestion
    let genericArguments: [TypeSyntax] = genericArgumentList?.arguments.map { $0.argument } ?? []
```